### ISSUE TITLE
Distributed checkpoint state_dict load may report nccl error and hide the real root-cause exception

### ISSUE THREAD

**Issue by bigning** [pytorch/pytorch#122529]
### 🐛 Describe the bug

Hi, when we call `torch.distributed.checkpoint.load_state_dict(state_dict, storage_reader)`, and if there is exception in storage_reader.read_data(), the exception is not immediately raised.[ It's captured here, and then dist.all_gather_objects(exception) before raising. ](https://github.com/pytorch/pytorch/blob/main/torch/distributed/checkpoint/utils.py#L246-L255)

Issue is:
1. [that `self.all_gather_object()`](https://github.com/pytorch/pytorch/blob/main/torch/distributed/checkpoint/utils.py#L251) might timeout if only one rank failed storage_reader.read_data() and the remaining ranks continue read_data, which takes longer time than nccl timeout. 
2. User might use a customized `storage_reader` (e.g. downloading from cloud), and there might be nccl communication in the `customized_storage_reader.read_data()`. In this case, if one rank failed `customized_storage_reader.read_data()`, the failure rank starts this [all_gather_object()](https://github.com/pytorch/pytorch/blob/main/torch/distributed/checkpoint/utils.py#L251), the remaining ranks start the other nccl comms in the `customized_storage_reader.read_data()`. This causes nccl comms mismatch 100%. 

What's worse, in both above case, the program crashes caused by nccl error before raising the real `storage_reader.read_data()` exception. The real exception message will never be visible to user. 

To print the `read_data` exception message, we have to capture the exception and  immediately print the exception message inside `read_data`. Example: https://github.com/mosaicml/composer/pull/3131

Can we just raise the `read_data()` exception immediately for the failure rank? I understand that the non-failure rank will continue run and timeout eventually at the next nccl comm, but at least the failure rank can print the real root-cause. 

### Versions

PyTorch version: 2.3.0.dev20240110+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

**Comment by bigning** [comment]
cc @mvpatel2000

**Comment by wconstab** [comment]
IIUC the issue is that 
1) pytorch first 'swallows' an error related to storage_reader.read_data() call
2) pytorch attempts to communicate this error with peers, but the communication won't proceed becuase (1) is already blocking the comms channel with some failed comms
3) the user sees a nccl error rather than the original storage_reader.read_data() error.

The requested behavior change is to surface the error from (1) immediately.

Is it enough if we print the exception and then continue to 'communicate' it?  Or do you want us to also crash the program at the point of the exception in (1)?

I'm not sure if i understand the reason for (2).  Why do we need this generally?  

cc @LucasLLC @fegin

**Comment by bigning** [comment]
Hi @wconstab , thanks for looking at this. 

> Is it enough if we print the exception and then continue to 'communicate' it? Or do you want us to also crash the program at the point of the exception in (1)?

Not really. We did that in the beginning to just [print the the error](https://github.com/mosaicml/composer/pull/3131/files) before 'communicate it'.  But it's not good since:
1. for large model, we need to set the nccl time out to be ~30 minutes. Which means it takes ~30 mins to crash with nccl timeout. 
2. It might happen that only 1 rank has the read_data() error. Since the nccl timedout rank is not necessarily the read_data() error rank, we have to look at the logging for all the ranks to see if it's read_data() error. This is not good for training with thousands of GPUs. 

So in order to mitigate above two issues, we had to [add an extra nccl comms inside read_data() ](https://github.com/mosaicml/composer/pull/3345/files)to communicate the checkpoint downloading error, and crash the program immediately  :(

**Comment by LucasLLC** [comment]
I think the suggestion here is pretty reasonable -- currently I believe the motivation here is to propagate the error to rank0 for easier triaging, etc., but if this behavior is hiding the error then we should reconsider.

**Comment by bigning** [comment]
@LucasLLC , @wconstab , created PR to run `read_data()` without all_gather()

**Comment by LucasLLC** [comment]
@bigning , it feels risky to completely remove the collective call on the load path, which at least in most cases:

1. does correctly propagate the error to rank0 (please correct me if I'm wrong)
2. Provides a synchronization point which signifies loading is complete.

If 1 is not the case (or we consider it unhelpful), then we should probably at least add a barrier for 2. But again, completely removing the mechanism might be considered BC breaking for users which are relying on errors to be broadcast.

Any thoughts on this @fegin ?

**Comment by bigning** [comment]
> @bigning , it feels risky to completely remove the collective call on the load path, which at least in most cases:
> 
> 1. does correctly propagate the error to rank0 (please correct me if I'm wrong)
> 2. Provides a synchronization point which signifies loading is complete.
> 
> If 1 is not the case (or we consider it unhelpful), then we should probably at least add a barrier for 2. But again, completely removing the mechanism might be considered BC breaking for users which are relying on errors to be broadcast.
> 
> Any thoughts on this @fegin ?

2 is a good point. Added dist.barrier()

**Comment by LucasLLC** [comment]
@bignig do you have a small repro for 1?

**Comment by bigning** [comment]
i don't have a small repro right now, but it should be something like this:
two gpus, in read_data, adding a nccl comms, and make rank-0 raise exception before that nccl comms. So that rank-0 will run the all_gather to try to communicate the exception, but rank-1 will run the nccl comms inside the read_data. 

@LucasLLC , does that make sense ?

**Comment by LucasLLC** [comment]
@bignig makes sense, thanks. Could you explain a little more why we need collectives in `read_data`? I'm not sure we designed for this

**Comment by bigning** [comment]
> Could you explain a little more why we need collectives in read_data? 

Good question! It's again the remote checkpoint downloading from cloud. 

**background**: when we download remote checkpoint files from cloud, we want to download as few files as possible because:
1. to save cost of network traffic
2. Downloading from cloud is not stable/reliable. They more files we download, the more likely it fails. 

**In the below two scenarios:** 
1. **FSDP**: Let's say for FSDP, the saved checkpoints is from a job with world_size x, but when loading, we use 2x world_size, so it's possible that rank_0 and rank_1 needs the same remote file. since rank_0 and rank_1 are on the same node, we don't need to download the same file on both ranks. We only want to rank_0 to download the file.
2. **HSDP**. In HSDP case, we only want 1 replica to download the files, then broadcast to other replicas. 

In both scenarios, we need nccl comms to coordinate the downloading or communicating the checkpoint files. [Here is our read_data implementation](https://github.com/mosaicml/composer/blob/dfbaf1436453db6b186d75e6ec50c222e8a8b44b/composer/utils/checkpoint.py#L237) for a remoteStorageReader

**Comment by fegin** [comment]
@wconstab What's the recommended communication mechanism for users to implement control plan? NCCL or GLOO?

**Comment by LucasLLC** [comment]
I've been thinking about this, and I believe even a barrier cause this conflict as well since it's almost equivalent to the allreduce. If users are using collectives in `read_data`, the suggestion might just be to use a separate process group.

**Comment by bigning** [comment]
>  I believe even a barrier cause this conflict 

@LucasLLC , but it's much better than the all_gather(exceptioin). I think the key point here is not barrier vs other comms. It's that we should avoid swallow the exception in read_data. With `dist.barrier`, the exception rank should crash immediately and print the stack trace. Other ranks might still need to wait nccl timeout. It's better than now, where all rank nccl timeout and no ranks prints the correct stacktrace.
